From edb9365c73663a3ec9b00e5df02888724c5ff404 Mon Sep 17 00:00:00 2001
From: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
Date: Fri, 23 Jun 2023 15:58:51 +0200
Subject: [PATCH 51/87] printk: Check only for migration in
 printk_deferred_*().

Atomic context is not required by the implementation. The only
requirement is that the caller does not migrate to another CPU between
the _enter() and _exit() invocation. The reason is to increment and
decrement the per-CPU variable on the same CPU.
Checking for migration only allows to use deferred printk on PREEMPT_RT
when only sleeping locks are acquired.

Check for disabled migration instead for atomic context in
printk_deferred_*()

Signed-off-by: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
---
 kernel/printk/printk_safe.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/kernel/printk/printk_safe.c b/kernel/printk/printk_safe.c
index 5c1470bd60bc..a324eaeb2533 100644
--- a/kernel/printk/printk_safe.c
+++ b/kernel/printk/printk_safe.c
@@ -39,13 +39,13 @@ void __printk_safe_exit(unsigned long *flags)
 
 void __printk_deferred_enter(void)
 {
-	WARN_ON_ONCE(!in_atomic());
+	cant_migrate();
 	this_cpu_inc(printk_context.recursion);
 }
 
 void __printk_deferred_exit(void)
 {
-	WARN_ON_ONCE(!in_atomic());
+	cant_migrate();
 	this_cpu_dec(printk_context.recursion);
 }
 
-- 
2.34.1

